api/src/main/java/org/apache/iceberg/util/StructProjection.java

```diff
@@ -189,16 +189,12 @@ public int size() {
 
   @Override
   public <T> T get(int pos, Class<T> javaClass) {
-    if (struct == null) {
-      // Return a null struct when projecting a nested required field from an optional struct.
-      // See more details in issue #2738.
-      return null;
-    }
-
     int structPos = positionMap[pos];
-
     if (nestedProjections[pos] != null) {
-      return javaClass.cast(nestedProjections[pos].wrap(struct.get(structPos, StructLike.class)));
+      StructLike nestedStruct = struct.get(structPos, StructLike.class);
```

Comment:
is there a potential npe here if "struct" is null?

Reply:
good question. this change avoided the wrapping null at nested level, [which is the main reason we added the lines from 192-196](https://github.com/apache/iceberg/issues/2738) that are removed in this PR. 

I guess it can potentially happen if we use `StructProjection` to wrap a null `struct` at the root level. If it is a valid scenario, we need to bring back line 192-196. If it is not a valid scenario, we should add `Preconditions` check in the wrap method.
```
  public StructProjection wrap(StructLike newStruct) {
    this.struct = newStruct;
    return this;
  }
```

Reply:
Even if we need to bring back line 192-196, it may still be good to return null projection 
```
Struct(1, null)
->
before this change: StructProjection(1, StructProjection(null))
after this change:    StructProjection(1, null)
```

Reply:
I agree about the behavior to return nested nulls without wrapping but I am not sure it is a good idea to remove lines 192-196. This API is widely used and it does not seem like a totally weird use case to wrap top-level nulls. Otherwise, every caller would have to check if the passed value is null. If someone has to do that check, better hide it inside this class.

Reply:
> This API is widely used and it does not seem like a totally weird use case to wrap top-level nulls.

that is a fair point. will bring back the null check. but will

Reply:
@RussellSpitzer @aokolnychyi I have a second thought. The behaviors of wrapping null root value seems a little odd. Note that `StructProjection` is also a `StructLike`.

```
struct = null
projection = StructProjection(null)
```

`struct == null` would be true. but `projection == null` would be false. `StructProjection(null)` behaves more like an empty struct.

also note that line 192-196 were originally added to deal with nested null struct (issue #2738 and PR #), which is avoided in the PR #3240.

Reply:
After a closer look, you are probably right. Let me check usages but seems reasonable to remove that.

Reply:
I don't see any blockers. Let's remove those lines and get this in.

Reply:
I agree with this change. The only issue is what happens when struct is null. That could still happen if `wrap` is never called.

I think I would favor the original null check (rather than the precondition) and the new behavior to return null instead of a `StructProjection` that wraps null. That way if somehow a null slips in, we at least just produce null. That seems to fit the intent as well, which is to return `null` when there is no value.

Reply:
> The only issue is what happens when struct is null. That could still happen if wrap is never called.

This is a good point. will revert to the original null check then.